b9lab-content/5-my-own-chain/03-starport-09-game-fifo.md

```diff
@@ -0,0 +1,191 @@
+
+## The How
+
+You need another data structure. The simplest one is a FIFO, First-In-First-Out, that is constantly updated so that:
+
+* The games that have just been played on are taken out of wherever they are and sent to the tail.
+* The games that have not been played on the longest eventually end up at the head.
+
+So when terminating expired games in `EndBlock`, you keep taking expired games that are at the head of the FIFO. And stop until the head has an on-going game. The cost is therefore:
+
+* `O(1)` on each game creation and game play.
+* `O(k)` where `k` is the number of expired games on each block. And of course `k <= n`, normally by a lot.
+
+Isn't `k` still an unbounded number of operations? Yes, but if you use the same expiration duration on each game, for `k` games to expire together in a block, these `k` games would all have had to see an action in the same previous one block. Give or take the block before or after. So in the worst case, the largest `EndBlock` computation will be proportional to the largest regular block in the past. This is a reasonable risk to take. And remember this back-of-the-envelope assessment works only if the expiration duration is the same for all games, instead of being a parameter left to the potentially malicious game creator.
+
+## New Information
+
+How do you implement a FIFO given all the previous work? Let's choose a doubly linked list for that.
+
+1. You need to keep the id of the head and that of the tail. A ready-made place for that is `NextGame`. Good thing that it is already an object and is expandable. So in terms of code, just add a bit in `proto/checkers/next_game.proto`:
+    ```proto [https://github.com/cosmos/b9-checkers-academy-draft/blob/2343af69cd1f2c22acfac13f46393aa8ce686685/proto/checkers/next_game.proto#L11-L12]
+    message NextGame {
+        ...
+        string fifoHead = 3; // Will contain the index of the game at the head.
+        string fifoTail = 4; // Will contain the index of the game at the tail.
+    }
+    ```
+2. Each game needs to know which game is before it, and which is after it. The right place for this information is `StoredGame` itself. In `proto/checkers/stored_game.proto`:
+    ```proto [https://github.com/cosmos/b9-checkers-academy-draft/blob/2343af69cd1f2c22acfac13f46393aa8ce686685/proto/checkers/stored_game.proto#L16-L17]
+    message StoredGame {
+        ...
+        string beforeId = 8; // Pertains to the FIFO. Towards head.
+        string afterId = 9; // Pertains to the FIFO. Towards tail.
+    }
+    ```
+3. Finally, there needs to be an id that indicates _no game_. Let's use `"-1"`. In `x/checkers/types/keys.go `:
+    ```go [https://github.com/cosmos/b9-checkers-academy-draft/blob/2343af69cd1f2c22acfac13f46393aa8ce686685/x/checkers/types/keys.go#L32-L34]
+    const (
+        NoFifoIdKey = "-1"
+    )
+    ```
+
+To have Starport and Protobuf recompile these files, you can use:
```

Comment:
If the intention is to build and compile, not to actually start running the chain, consider using

`starport chain build`
https://docs.starport.network/cli/#starport-chain-build